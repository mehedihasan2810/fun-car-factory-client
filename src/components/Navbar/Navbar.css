.nav-container {
  position: fixed;
  top: 0;
  left: 0;
  z-index: 999;
  width: 100vw;
  background-color: rgba(0, 0, 0, 0.5);
  border-bottom: 1px solid rgba(255, 255, 255, 0.1);
  transform: translateY(0);
  transition: transform 300ms ease-in;
}

@supports (backdrop-filter: blur(10px)) {
  .nav-container {
    background-color: unset;
    backdrop-filter: blur(10px);
    -webkit-backdrop-filter: blur(10px);
  }
}

nav {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding-inline: 1rem;
  padding-block: 0.8rem;
  width: min(100vw, 1600px);
  margin-inline: auto;
}

nav img {
  width: 55px;
  height: 55px;
  object-fit: contain;
  filter: brightness(2);
}

nav .logo-container {
  display: flex;
  align-items: center;
  gap: 0.5rem;
}
nav ul {
  display: flex;
  align-items: center;
  gap: 1.5rem;
}

nav .btn-primary {
  font-size: 1.1rem;
  box-shadow: -2px 2px 0px var(--deep-white-100);
}

nav li {
  color: var(--deep-white-100);
}

nav li:hover a {
  color: var(--primary-sky-400);
  text-decoration: underline;
}

nav .profile-img {
  width: 40px;
  height: 40px;
  border-radius: 50%;
  object-fit: contain;
}

nav .active {
  color: var(--primary-sky-400);
  text-decoration: underline;
}
nav .pending {
  opacity: 0.5;
}

nav .bar {
  color: #fff;
  display: none;
  width: 28px;
  height: 28px;
}

nav .bar > * {
  width: 100%;
  height: 100%;
}

.nav-favorite-list svg {
  width: 30px;
  height: 30px;
  color: var(--deep-white-100);
}

.nav-favorite-list svg:hover {
  color: var(--base-gray-400);
}

@media (max-width: 768px) {
  nav {
    padding-block: 0.5rem;
  }
  nav .bar {
    display: block;
    background-color: transparent;
    border: none;
    font-size: 1.7rem;
    cursor: pointer;
  }
  nav img {
    width: 50px;
    height: 50px;
  }
  nav ul {
    display: none;
  }

  nav ul.open {
    display: block;
    position: fixed;
    top: 70px;
    right: 10px;
    background-color: var(--primary-sky-900);
    display: flex;
    flex-direction: column;
    padding: 2rem 4rem;
    border-radius: 4px;
    border: 1px solid rgba(255, 255, 255, 0.15);
  }
}
