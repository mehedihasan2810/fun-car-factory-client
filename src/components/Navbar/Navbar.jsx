import { NavLink, useNavigate } from "react-router-dom";
import { FaUserTie, FaBars } from "react-icons/fa";
import { HiOutlineX } from "react-icons/hi";
import { toast } from "react-toastify";
import "./Navbar.css";
import { useLayoutEffect, useRef, useState } from "react";
import gsap from "gsap";
import { ScrollTrigger } from "gsap/ScrollTrigger";
import { useAuthContext } from "../../contexts/useAuthContext";
import useContextProvider from "../../contexts/useContextProvider";
import Cookies from "js-cookie";
import { apolloClient } from "../../lib/graphql";

gsap.registerPlugin(ScrollTrigger);

const Navbar = () => {
  const navContainerRef = useRef();
  const [isOpenNav, setIsOpenNav] = useState(false);

  const { totalCartToys } = useContextProvider();
  // checkTotalCartToys();
  const { currentUser, logOut } = useAuthContext();
  const navigate = useNavigate();

  const handleSignOut = async () => {
    try {
      await logOut();

      Cookies.remove("token");

      apolloClient.resetStore();

      toast.success("Succesfully Signed Out", {
        position: toast.POSITION.TOP_CENTER,
        autoClose: 2000,
      });

      navigate("/");
    } catch (error) {
      // *show toast
      toast.error(error.message, {
        position: toast.POSITION.TOP_CENTER,
        autoClose: 2000,
      });
    }

    // logOut()
    //   .then(() => {
    //     // *show toast
    //     if (currentUser) {
    //       toast.success("Succesfully Signed Out", {
    //         position: toast.POSITION.TOP_CENTER,
    //         autoClose: 2000,
    //       });

    //       navigate("/");
    //     }
    //   })
    //   .catch((error) => {
    //     // *show toast
    //     toast.error(error.message, {
    //       position: toast.POSITION.TOP_CENTER,
    //       autoClose: 2000,
    //     });
    //   });
  };

  useLayoutEffect(() => {
    const links = document.querySelectorAll("nav li");

    links.forEach((link) => {
      link.addEventListener("click", () => {
        setIsOpenNav(false);
      });
    });

    ScrollTrigger.observe({
      target: window,
      type: "wheel,touch,scroll,pointer",
      onUp: () => {
        navContainerRef.current.style.transform = "translateY(0%)";
      },
      onDown: () => {
        navContainerRef.current.style.transform = "translateY(-100%)";
        setIsOpenNav(false);
      },
    });
  }, []);

  return (
    <div ref={navContainerRef} className="nav-container">
      <nav>
        <div className="logo-container">
          <img src="/assets/logo.png" alt="logo" width={55} height={55} />
        </div>
        <ul className={isOpenNav ? "open" : ""}>
          {/* {currentUser && <li data-testid="foo">Foooo</li>} */}

          <li>
            <NavLink
              to="/"
              className={({ isActive, isPending }) =>
                isPending ? "pending" : isActive ? "active" : ""
              }
            >
              Home
            </NavLink>
          </li>
          <li>
            <NavLink
              to="/all-toys"
              className={({ isActive, isPending }) =>
                isPending ? "pending" : isActive ? "active" : ""
              }
            >
              All Toys
            </NavLink>
          </li>
          <li>
            <NavLink
              to="/my-toys"
              className={({ isActive, isPending }) =>
                isPending ? "pending" : isActive ? "active" : ""
              }
            >
              My Toys
            </NavLink>
          </li>
          <li>
            <NavLink
              to="/add-toy"
              className={({ isActive, isPending }) =>
                isPending ? "pending" : isActive ? "active" : ""
              }
            >
              Add Toy
            </NavLink>
          </li>
          {currentUser ? (
            <>
              <li>
                {currentUser?.photoURL ? (
                  <img
                    data-testid="nav-user-img"
                    title={currentUser?.displayName}
                    className="profile-img"
                    src={currentUser?.photoURL}
                    alt="user photo"
                    width={40}
                    height={40}
                  />
                ) : (
                  <div title="No Name">
                    <FaUserTie style={{ width: 40, height: 40 }} />
                  </div>
                )}
              </li>
              <li>
                <button
                  data-testid="signOut"
                  onClick={handleSignOut}
                  className="btn-primary"
                >
                  Sign Out
                </button>
              </li>
            </>
          ) : (
            <li>
              <NavLink to="/signin">
                <button className="btn-primary">Sign In</button>
              </NavLink>
            </li>
          )}

          <li className="nav-favorite-list">
            <NavLink to="/favorites">
              <svg
                xmlns="http://www.w3.org/2000/svg"
                fill="none"
                viewBox="0 0 24 24"
                strokeWidth={1.5}
                stroke="currentColor"
              >
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  d="M21 8.25c0-2.485-2.099-4.5-4.688-4.5-1.935 0-3.597 1.126-4.312 2.733-.715-1.607-2.377-2.733-4.313-2.733C5.1 3.75 3 5.765 3 8.25c0 7.22 9 12 9 12s9-4.78 9-12z"
                />
              </svg>
            </NavLink>
          </li>
          <li className="nav-cart-list">
            <NavLink to="/cart">
              <svg
                xmlns="http://www.w3.org/2000/svg"
                fill="none"
                viewBox="0 0 24 24"
                strokeWidth={1.5}
                stroke="currentColor"
                className="w-6 h-6"
              >
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  d="M2.25 3h1.386c.51 0 .955.343 1.087.835l.383 1.437M7.5 14.25a3 3 0 00-3 3h15.75m-12.75-3h11.218c1.121-2.3 2.1-4.684 2.924-7.138a60.114 60.114 0 00-16.536-1.84M7.5 14.25L5.106 5.272M6 20.25a.75.75 0 11-1.5 0 .75.75 0 011.5 0zm12.75 0a.75.75 0 11-1.5 0 .75.75 0 011.5 0z"
                />
              </svg>
              <div className="nav-cart-count">{totalCartToys}</div>
            </NavLink>
          </li>
        </ul>
        <button
          onClick={() => {
            setIsOpenNav(!isOpenNav);
          }}
          className="bar"
        >
          {isOpenNav ? <HiOutlineX /> : <FaBars />}
        </button>
      </nav>
    </div>
  );
};

export default Navbar;
