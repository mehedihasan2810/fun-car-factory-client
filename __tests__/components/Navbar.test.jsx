import {
  customRender,
  describe,
  expect,
  screen,
  test,
} from "../utils/utils";
import Navbar from "../../src/components/Navbar/Navbar";

describe("<Navbar />", () => {
  test("Should render succesfully", async () => {
    const { asFragment } = customRender(<Navbar />);

    // const { result } = renderHook(useAuthContext, {
    //   wrapper: customWrapper,
    // });

    const button1 = screen.getByText("Home");
    const button2 = screen.getByText("All Toys");
    expect(button1).toBeInTheDocument();
    expect(button2).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
