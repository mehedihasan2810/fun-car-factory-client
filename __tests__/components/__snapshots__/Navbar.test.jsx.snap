// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Navbar /> > Should render succesfully 1`] = `
<DocumentFragment>
  <div
    class="nav-container"
  >
    <nav>
      <div
        class="logo-container"
      >
        <img
          alt="logo"
          height="55"
          src="/assets/logo.png"
          width="55"
        />
      </div>
      <ul
        class=""
      >
        <li>
          <a
            aria-current="page"
            class="active"
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            class=""
            href="/all-toys"
          >
            All Toys
          </a>
        </li>
        <li>
          <a
            class=""
            href="/my-toys"
          >
            My Toys
          </a>
        </li>
        <li>
          <a
            class=""
            href="/add-toy"
          >
            Add Toy
          </a>
        </li>
        <li>
          <a
            class=""
            href="/signin"
          >
            <button
              class="btn-primary"
            >
              Sign In
            </button>
          </a>
        </li>
        <li
          class="nav-favorite-list"
        >
          <a
            class=""
            href="/favorites"
          >
            <svg
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 8.25c0-2.485-2.099-4.5-4.688-4.5-1.935 0-3.597 1.126-4.312 2.733-.715-1.607-2.377-2.733-4.313-2.733C5.1 3.75 3 5.765 3 8.25c0 7.22 9 12 9 12s9-4.78 9-12z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </a>
        </li>
        <li
          class="nav-cart-list"
        >
          <a
            class=""
            href="/cart"
          >
            <svg
              class="w-6 h-6"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2.25 3h1.386c.51 0 .955.343 1.087.835l.383 1.437M7.5 14.25a3 3 0 00-3 3h15.75m-12.75-3h11.218c1.121-2.3 2.1-4.684 2.924-7.138a60.114 60.114 0 00-16.536-1.84M7.5 14.25L5.106 5.272M6 20.25a.75.75 0 11-1.5 0 .75.75 0 011.5 0zm12.75 0a.75.75 0 11-1.5 0 .75.75 0 011.5 0z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            <div
              class="nav-cart-count"
            >
              0
            </div>
          </a>
        </li>
      </ul>
      <button
        class="bar"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
          />
        </svg>
      </button>
    </nav>
  </div>
</DocumentFragment>
`;
